README.md
# Feathers Starter Kit & Vuetify or Framework7-Vue Client 

> app dev made easy


## Install prerequisites on dev machine and production server
- tested with ubuntu 16.04 and node 8.x or 10.x

### Install node.js(8.x or 10.x) and git
``` sh
# Install dependencies on ubuntu
curl -sL https://deb.nodesource.com/setup_8.x | sudo -E bash -
# Install node and git
sudo apt-get install git nodejs
```

## First steps
#### Clone repos and install deps
```
git clone https://github.com/akralj/app {{yourAppName}}
cd {{yourAppName}}

### Install dependencies
npm install
```

#### Configure app
- change name, description & author in ./package.json. Name will be used in deploy script (scripts/uploadAndRestartApp.sh)
- change appPort to production port,... in ./server/config.coffee to match your enviroment

#### Start with a fresh git repo
```
rm -rf .git
git init
git add .
git commit -am "initial commit"
# Add your own remote origin
git remote add origin ssh://git@YourSourcecodeServer/projectName/appName.git
git push -u origin master
```

## Prepare Production Server
- see docs/deploy.md


# Note:
There are 2 package.json files in this project, in "./" and "./client/"

1. "./package.json" is used for server dependencies (dependencies), build tools and npm scripts (devDependencies)
2. "./client/package.json" is used for the client dependencies

### IMPORTANT
- Please install client dependencies in ../client


# Development Workflow

run dev server on port 7777. Same port as later in production.
```
npm run dev
```

build and run "production" code on staging server on devServer:7778
```
npm run stage
```

# Production Workflow
### Step 0
change ./package.json version: to new release, eg. from release-0.1.1 to release-0.1.2

### Step 1

run development instance
```
npm run dev &
npm run stage
```
go to http://devServer:7778 # (not implemnted in parcel setup yet) and /app.html if it's an offline app
do manual tests from trello

### Step 2

create production app with the user context the service runs and you have ssh access.

```
DEPLOY_SSH=user@server npm run deploy
```
or configure DEPLOY_SSH in ./jobs/uploadAndRestartApp.sh

- builds client and copy client app to server/public
- rsync ./server, ./node_modules and ./package.json to production server (prodServerName)
- stop and start app-{{appName}} (systemd) on production server

### Step 3

check if service running
```
ssh username@prodServerName
sudo journalctl -u app-{{appName} -f
```

### Step 4

Commit to dev
```gca "release-x.x.x - your comment" ```

Commit to master
```
git co master
git merge dev
git push origin master
git co dev
```

TODOS: add more docs like in: http://vuejs-templates.github.io/webpack/


CHANGELOG.md
# Change Log

## app-name v0.2.0 - Updated on 04.07.2019
- switch to parcel bundler
- upgrade to vuetify 2.0


scripts/uploadAndRestartApp.sh
#!/bin/env bash
# DEPLOY_SSH="user@serverName" npm run deploy # e.g.  DEPLOY_SSH="a@diolapp02" npm run deploy

# 1. get name prop from package.json
APP_NAME="$(node -pe "require('./package.json')['name']")"
# you also can set it here
#DEPLOY_SSH="user@serverName"


if [ -n "$DEPLOY_SSH" ]
then
  # 2. copy server app to production
  rsync -avh --stats --delete server/       $DEPLOY_SSH:/apps/$APP_NAME/server
  rsync -avh --stats --delete node_modules/ $DEPLOY_SSH:/apps/$APP_NAME/node_modules
  rsync -avh --stats          package.json  $DEPLOY_SSH:/apps/$APP_NAME
  # 3. restart systemd service on production server
  ssh $DEPLOY_SSH "sudo systemctl stop app-$APP_NAME; sudo systemctl start app-$APP_NAME"

else
  echo "Please set DEPLOY_SSH=ssh path to server, e.g. DEPLOY_SSH=user@server.domain.at npm run deploy"
fi


client/webpack.config.js
var path = require('path')
var webpack = require('webpack')
var ExtractTextPlugin = require("extract-text-webpack-plugin")
var MinifyPlugin = require("babel-minify-webpack-plugin")


// get serverPort from main server/config.coffee
try {
  require('coffeescript').register();
  serverConfig = require("../server/config")("development");
  serverPort = serverConfig.appPort
} catch (error) {
  // default serverPort if serverConfig is missing or no server part in project
  // access app at: http://localhost:7777
  serverPort = 7778
}


// rules used in dev and production
var baseRules =  [
  {
    test: /\.(png|jpg|gif|svg)$/,
    loader: 'file-loader',
    options: {
      objectAssign: 'Object.assign'
    }
  },
  {
    test: /\.styl$/,
    loader: ['style-loader', 'css-loader', 'stylus-loader']
  },
  {
    test: /\.coffee$/,
    use: [ 'coffee-loader' ]
  }
]


module.exports = {
  entry: './src/index.coffee',
  output: {
    path: path.resolve(__dirname, './dist'),
    publicPath: '/dist/',
    filename: 'build.js'
  },
  resolve: {
    extensions: ['.coffee', '.js', '.vue', '.json'],
    alias: {
      'vue$': 'vue/dist/vue.esm.js',
      'static': path.resolve(__dirname, './static')
    }
  },

  module: {
    rules: baseRules.concat([
      {
        test: /\.vue$/,
        loader: 'vue-loader',
        options: {
          loaders: {
            coffee: "coffee-loader"
          }
        }
      }
    ])
  },
  devServer: {
    historyApiFallback: true,
    noInfo: true,
    disableHostCheck: true,
    host: "0.0.0.0",
    port: serverPort - 1 ,
    proxy: [
    {
      context: [ "/api", "/assets"],
      target: "http://0.0.0.0:" + serverPort
    }
  ]
  },
  performance: {
    hints: false
  },
  devtool: '#eval-source-map'
}

if (process.env.NODE_ENV === 'production') {

  module.exports.module.rules = baseRules.concat([
    {
      test: /\.vue$/,
      loader: 'vue-loader',
      options: {
        loaders: {
          css: ExtractTextPlugin.extract({
            use: 'css-loader',
            fallback: 'vue-style-loader' // <- this is a dep of vue-loader, so no need to explicitly install if using npm3
          })
        }
      }
    }
  ])

  module.exports.devtool = '#cheap-source-map'
  // http://vue-loader.vuejs.org/en/workflow/production.html
  module.exports.plugins = (module.exports.plugins || []).concat([
    new webpack.DefinePlugin({
      'process.env': {
        NODE_ENV: '"production"'
      }
    }),
    new ExtractTextPlugin("style.css"),
    // coffeescript1, if you need to support older browsers
    //new webpack.optimize.UglifyJsPlugin({ sourceMap: true, compress: { warnings: false } })
    new MinifyPlugin() // for coffeescript2
  ])
}
